File: processing/syntactic/expressions/definitions/module_def.py
from __future__ import annotations

from typing import List

from processing.codegen.code_builder import CodeBuilder
from processing.lexical.tokens.token_type import TokenType
from processing.syntactic.expressions.atomic.name_expr import NameExpr
from processing.syntactic.expressions.block_expr import BlockExpr, BlockType
from processing.syntactic.expressions.expr import Expr, child_property
from processing.syntactic.expressions.expression_groups import DefinitionExpression


class ModuleDef(DefinitionExpression):
    """module_def:
       'module' name block;
    """

    @child_property
    def name(self) -> NameExpr:
        pass

    @child_property
    def block(self) -> BlockExpr:
        pass

    @child_property
    def modifiers(self) -> List[Expr]:
        pass

    def __init__(
            self,
            parent: Expr = None,
            name: NameExpr = None,
            block: BlockExpr = None,
            modifiers: List[Expr] = None
    ):
        super().__init__(parent)
        self.name = name
        self.block = block
        self.modifiers = modifiers

    def parse(self) -> ModuleDef:
        self.stream.eat(TokenType.keyword_module)
        self.name = NameExpr(self).parse(must_be_simple = True)
        self.block = BlockExpr(self).parse(BlockType.default)
        return self

    def to_axion(self, c: CodeBuilder):
        c += 'module ', self.name, self.block

    def to_csharp(self, c: CodeBuilder):
        c += 'namespace ', self.name, self.block

File: Axion.Core/Processing/Syntactic/Expressions/Definitions/ModuleDef.cs
using System.Collections.Generic;
using Axion.Core.Processing.Lexical.Tokens;
using Axion.Core.Processing.Syntactic.Expressions.Atomic;
using Magnolia.Attributes;
using Magnolia.Trees;
using static Axion.Specification.TokenType;

namespace Axion.Core.Processing.Syntactic.Expressions.Definitions;

/// <summary>
///     <code>
///         module-def:
///             'module' name scope;
///     </code>
/// </summary>
[Branch]
public partial class ModuleDef : Node, IDefinitionExpr, IDecorableExpr {
    [Leaf] Token? kwModule;
    [Leaf] NameExpr? name;
    [Leaf] ScopeExpr scope = null!;

    public ModuleDef(Node parent) : base(parent) { }

    public DecoratedExpr WithDecorators(params Node[] items) {
        return new(Parent) {
            Target     = this,
            Decorators = new NodeList<Node, Ast>(this, items)
        };
    }

    public ModuleDef WithScope(params Node[] items) {
        return WithScope((IEnumerable<Node>) items);
    }

    public ModuleDef WithScope(IEnumerable<Node> items) {
        Scope = new ScopeExpr(this).WithItems(items);
        return this;
    }

    public ModuleDef Parse() {
        KwModule = Stream.Eat(KeywordModule);
        Name     = new NameExpr(this).Parse();
        Scope    = new ScopeExpr(this).Parse();
        return this;
    }
}
